samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.zk;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.I0Itec.zkclient.IZkDataListener;
+import org.apache.samza.SamzaException;
+import org.apache.samza.coordinator.leaderelection.LeaderElector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.Random;
+
+/**
+ * <p>
+ * An implementation of Leader Elector using Zookeeper.
+ *
+ * Each participant in the leader election process creates an instance of this class and tries to become the leader.
+ * The participant with the lowest sequence number in the ZK subtree for election becomes the leader. Every non-leader
+ * sets a watcher on its predecessor, where the predecessor is the participant with the largest sequence number
+ * that is less than the current participant's sequence number.
+ * </p>
+ * */
+public class ZkLeaderElector implements LeaderElector {
+  public static final Logger LOGGER = LoggerFactory.getLogger(ZkLeaderElector.class);
+  private final ZkUtils zkUtils;
+  private final String processorIdStr;
+  private final ZkKeyBuilder keyBuilder;
+  private final String hostName;
+
+  private String leaderId = null;
+  private final IZkDataListener zkLeaderElectionListener;
+  private String currentSubscription = null;
+  private final Random random = new Random();
+
+  @VisibleForTesting
+  ZkLeaderElector(String processorIdStr, ZkUtils zkUtils, IZkDataListener leaderElectionListener) {
+    this.processorIdStr = processorIdStr;
+    this.zkUtils = zkUtils;
+    this.zkLeaderElectionListener = leaderElectionListener;
+    this.keyBuilder = this.zkUtils.getKeyBuilder();
+    this.hostName = getHostName();
+  }
+
+  public ZkLeaderElector(String processorIdStr, ZkUtils zkUtils) {
+    this.zkLeaderElectionListener = new ZkLeaderElectionListener();
+    this.processorIdStr = processorIdStr;
+    this.zkUtils = zkUtils;
+    this.keyBuilder = this.zkUtils.getKeyBuilder();
+    this.hostName = getHostName();
+  }
+
+  // TODO: This should go away once we integrate with Zk based Job Coordinator
```

Comment:
Yeah. It has to be more than just the host name. It will be a host identifier structure (json data). I am still working on what the contents of that data has to be.